src/coordination/coordinator_cluster_state.cpp

```diff
@@ -0,0 +1,124 @@
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+#ifdef MG_ENTERPRISE
+
+#include "nuraft/coordinator_cluster_state.hpp"
+
+namespace memgraph::coordination {
+
+auto CoordinatorClusterState::MainExists() const -> bool {
+  return std::any_of(instance_roles.begin(), instance_roles.end(), [](auto const &entry) {
+    return entry.second == replication_coordination_glue::ReplicationRole::MAIN;
+  });
+}
+
+auto CoordinatorClusterState::IsMain(std::string const &instance_name) const -> bool {
+  auto const it = instance_roles.find(instance_name);
+  return it != instance_roles.end() && it->second == replication_coordination_glue::ReplicationRole::MAIN;
+}
+
+auto CoordinatorClusterState::IsReplica(std::string const &instance_name) const -> bool {
+  auto const it = instance_roles.find(instance_name);
+  return it != instance_roles.end() && it->second == replication_coordination_glue::ReplicationRole::REPLICA;
+}
+
+auto CoordinatorClusterState::InsertInstance(std::string const &instance_name,
+                                             replication_coordination_glue::ReplicationRole role) -> void {
+  instance_roles[instance_name] = role;
+}
+
+auto CoordinatorClusterState::DoAction(std::string const &instance_name, RaftLogAction log_action) -> void {
+  switch (log_action) {
+    case RaftLogAction::REGISTER_REPLICATION_INSTANCE:
+      instance_roles[instance_name] = replication_coordination_glue::ReplicationRole::REPLICA;
+      break;
+    case RaftLogAction::UNREGISTER_REPLICATION_INSTANCE:
+      spdlog::info("Unregistering instance: {}", instance_name);
+      instance_roles.erase(instance_name);
+      spdlog::info("Instance {} unregistered", instance_name);
+      break;
+    case RaftLogAction::SET_INSTANCE_AS_MAIN:
+      MG_ASSERT(instance_roles.find(instance_name) != instance_roles.end(),
+                "Instance does not exist as part of raft state!");
+      instance_roles[instance_name] = replication_coordination_glue::ReplicationRole::MAIN;
+      break;
+    case RaftLogAction::SET_INSTANCE_AS_REPLICA:
+      MG_ASSERT(instance_roles.find(instance_name) != instance_roles.end(),
+                "Instance does not exist as part of raft state!");
+      instance_roles[instance_name] = replication_coordination_glue::ReplicationRole::REPLICA;
+      break;
+  }
+}
+
+auto CoordinatorClusterState::Serialize(ptr<buffer> &data) -> void {
+  auto const role_to_string = [](auto const &role) {
+    switch (role) {
+      case replication_coordination_glue::ReplicationRole::MAIN:
+        return "main";
+      case replication_coordination_glue::ReplicationRole::REPLICA:
+        return "replica";
+    }
+  };
+
+  auto const entry_to_string = [&role_to_string](auto const &entry) {
+    return entry.first + "_" + role_to_string(entry.second);
+  };
+
+  auto instances_str_view = instance_roles | ranges::views::transform(entry_to_string);
+  uint32_t size =
```

Comment:
2 passes, this first pass does not need strings, you can calculate the would be size without doing `entry_to_string`

Reply:
how exactly? + ranges should just be lazy I think right? So at the end there should be just one pass for accumulate

Reply:
You run the range twice...once to calculate size (which does unnecessary temporary strings). And a 2nd time to put into the buffer.